offload_vio/server_rx/video_decoder.cpp

```diff
@@ -0,0 +1,162 @@
+
+    auto identity = gst_element_factory_make("identity", "identity");
+    g_object_set(G_OBJECT(identity), "dump", TRUE, nullptr);
+    g_object_set(G_OBJECT(identity), "signal-handoffs", TRUE, nullptr);
+
+    gst_bin_add_many(GST_BIN(_pipeline_img0), _appsrc_img0, h265parse_0, identity, decoder_img0, videoconvert_0,
+                     nvvideoconvert_0, caps_filter_0, _appsink_img0, nullptr);
+    gst_bin_add_many(GST_BIN(_pipeline_img1), _appsrc_img1, h265parse_1, decoder_img1, videoconvert_1, nvvideoconvert_1,
+                     caps_filter_1, _appsink_img1, nullptr);
+
+    // link elements
+    if (!gst_element_link_many(_appsrc_img0, decoder_img0, nvvideoconvert_0, videoconvert_0, caps_filter_0, _appsink_img0,
+                               nullptr) ||
+        !gst_element_link_many(_appsrc_img1, decoder_img1, nvvideoconvert_1, videoconvert_1, caps_filter_1, _appsink_img1,
+                               nullptr)) {
+        abort("Failed to link elements");
+    }
+
+    gst_element_set_state(_pipeline_img0, GST_STATE_PLAYING);
+    gst_element_set_state(_pipeline_img1, GST_STATE_PLAYING);
+}
+
+video_decoder::video_decoder(std::function<void(cv::Mat&&, cv::Mat&&)> callback)
+    : _callback(std::move(callback)) { }
+
+void video_decoder::init() {
+    create_pipelines();
+}
+
+void video_decoder::enqueue(std::string& img0, std::string& img1) {
+    auto buffer_img0 =
+        gst_buffer_new_wrapped_full(GST_MEMORY_FLAG_READONLY, img0.data(), img0.size(), 0, img0.size(), nullptr, nullptr);
+    auto buffer_img1 =
+        gst_buffer_new_wrapped_full(GST_MEMORY_FLAG_READONLY, img1.data(), img1.size(), 0, img1.size(), nullptr, nullptr);
+
+    GST_BUFFER_OFFSET(buffer_img0) = _num_samples;
+    GST_BUFFER_OFFSET(buffer_img1) = _num_samples;
+
+    _num_samples++;
+
+    auto ret_img0 = gst_app_src_push_buffer(reinterpret_cast<GstAppSrc*>(_appsrc_img0), buffer_img0);
+    auto ret_img1 = gst_app_src_push_buffer(reinterpret_cast<GstAppSrc*>(_appsrc_img1), buffer_img1);
+
+    if (ret_img0 != GST_FLOW_OK || ret_img1 != GST_FLOW_OK) {
+        abort("Failed to push buffer");
+    }
+}
+
+GstFlowReturn video_decoder::cb_appsink(GstElement* sink) {
+    // std::cout << "cb_appsink" << std::endl;
+
+    GstSample* sample;
+    g_signal_emit_by_name(sink, "pull-sample", &sample);
+    if (sample) {
+        GstBuffer* buffer = gst_sample_get_buffer(sample);
+
+        std::unique_lock<std::mutex> lock(_pipeline_sync_mutex); // lock acquired
+        if (sink == _appsink_img0) {
+            gst_buffer_map(buffer, &_img0_map, GST_MAP_READ);
+            _img0_ready = true;
+        } else {
+            gst_buffer_map(buffer, &_img1_map, GST_MAP_READ);
+            _img1_ready = true;
+        }
+
+        if (_img0_ready && _img1_ready) {
+            _callback(cv::Mat(IMG_HEIGHT, IMG_WIDTH, CV_8UC1, _img0_map.data),
+                      cv::Mat(IMG_HEIGHT, IMG_WIDTH, CV_8UC1, _img1_map.data));
```

Comment:
Yes it would be great to extend the encoder and decoder to support more image formats. In fact, I believe we have the plan to generalize offloading plugins as well as encoders and decoders, but that's out of scope of this PR.